espressostreamer/espresso_streamer.go

```diff
@@ -86,48 +86,118 @@ func NewEspressoStreamer(
-	defer s.messageMutex.Unlock()
 	s.currentMessagePos = currentMessagePos
 	s.nextHotshotBlockNum = currentHostshotBlock
 	s.messageWithMetadataAndPos = []*MessageWithMetadataAndPos{}
 }
 
-func (s *EspressoStreamer) Next() (*MessageWithMetadataAndPos, error) {
-	s.messageMutex.Lock()
-	defer s.messageMutex.Unlock()
+func (s *EspressoStreamer) Next(ctx context.Context) (*MessageWithMetadataAndPos, error) {
+	result, err := s.Peek(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-	message, found := FilterAndFind(&s.messageWithMetadataAndPos, func(msg *MessageWithMetadataAndPos) int {
+	// Advance the current message position, so that the next call to
+	// `Peek` or `Next` will return the next message
+	s.Advance()
+	return result, nil
+}
+
+func (s *EspressoStreamer) Peek(ctx context.Context) (*MessageWithMetadataAndPos, error) {
+	compareMessageWithCurrentPos := func(msg *MessageWithMetadataAndPos) int {
 		if msg.Pos == s.currentMessagePos {
-			return 0
+			return FilterAndFind_Target
 		}
 		if msg.Pos < s.currentMessagePos {
-			return -1
+			return FilterAndFind_Remove
+		}
+		return FilterAndFind_Keep
+	}
+
+	messageIndex := FilterAndFind(&s.messageWithMetadataAndPos, compareMessageWithCurrentPos)
+
+	if messageIndex >= 0 {
+		return s.messageWithMetadataAndPos[messageIndex], nil
+	}
+
+	condition := func(messages []*MessageWithMetadataAndPos) bool {
+		for _, message := range messages {
+			if message.Pos == s.currentMessagePos {
+				return true
+			}
 		}
-		return 1
-	})
-	if !found {
-		return nil, nil
+		return false
 	}
-	if message == nil {
-		// This should never happen.
-		return nil, fmt.Errorf("message is nil, but found is true")
+
+	// We set a timeout to avoid infinite loop in case of a bug.
+	timeout := 2 * time.Minute
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	err := s.QueueMessagesFromHotShotUntil(timeoutCtx, s.parseEspressoTransaction, condition)
+	if err != nil {
+		return nil, fmt.Errorf("failed to queue messages from hotshot until condition is met within %s", timeout)
 	}
 
+	messageIndex = FilterAndFind(&s.messageWithMetadataAndPos, compareMessageWithCurrentPos)
+
+	if messageIndex >= 0 {
+		return s.messageWithMetadataAndPos[messageIndex], nil
+	}
+	return nil, fmt.Errorf("message not found, please check the condition")
+}
+
+// Call this function to advance the streamer to the next message
+func (s *EspressoStreamer) Advance() {
 	s.currentMessagePos += 1
-	return message, nil
+}
+
+// This function keep fetching hotshot blocks and parsing them until the condition is met.
+// It is a do-while loop, which means it will always execute at least once.
+//
+// Expose the *parseHotShotPayloadFn* to the caller for testing purposes
+func (s *EspressoStreamer) QueueMessagesFromHotShotUntil(
+	ctx context.Context,
+	parseHotShotPayloadFn func(tx espressoTypes.Bytes) ([]*MessageWithMetadataAndPos, error),
+	condition func(messages []*MessageWithMetadataAndPos) bool,
+) error {
+	for {
```

Comment:
I think we need to make a call at the start of this function (before the for loop) that fetches the current espresso height and exits from the for loop if we reach the initial height.

That way we aren't making multiple calls to find the current espresso height. Additionally,  the streamer will be close enough to the tip of the chain at the end of the current call to `QueueMessagesFromHotshotUntil` that future calls to `Next`\`Peek` won't block for very long, and will be closer in the chain to where the message would likely reside if the current call doesn't find the message.